airflow/opt/providers/etna/etna/hooks/box.py

```diff
@@ -0,0 +1,237 @@
+
+        params:
+          folder_name: str, the top-level folder to search from.
+          batch_start, Optional[datetime], to compare against each file's mtime, via <=
+          batch_end, Optional[datetime], to compare against each file's mtime, via >=
+        """
+        if not Box.valid_folder_name(folder_name):
+            raise ValueError(f"Invalid folder name: {folder_name}. Only alphanumeric characters, _, -, and spaces are allowed.")
+
+        with self.ftps() as ftps:
+            all_files = self._ls_r(ftps)
+
+            return [f for f in all_files if self._is_in_range(f, batch_start, batch_end)]
+
+    def _ls_r(self, ftps: FTP_TLS, path: str = "/") -> List[FtpEntry]:
+        files = []
+        for entry in ftps.mlsd(path):
+            ftp_entry = FtpEntry(entry, path)
+
+            if ftp_entry.is_dot():
+                continue
+
+            if ftp_entry.is_dir():
+                files += self._ls_r(ftps, os.path.join(path, ftp_entry.name))
+            else:
+                files.append(ftp_entry)
+        return files
+
+    def _is_in_range(self, file: FtpEntry, batch_start: Optional[datetime] = None, batch_end: Optional[datetime] = None):
+        if batch_start is None and batch_end is None:
+            return True
+        elif batch_start is None:
+            return file.mtime <= batch_end
+        elif batch_end is None:
+            return file.mtime >= batch_start
+        else:
+            return batch_start <= file.mtime <= batch_end
+
+    @contextlib.contextmanager
+    def ftps(self) -> FTP_TLS:
+        """
+        Configures an FTP_TLS connection to Box. Using Pythong `with` syntax, so that the
+        connection is closed after usage.
+
+        eg:
+        ```
+        with box.ftps() as ftps:
+            ftps.sendcmd("LIST")
+        ```
+        """
+        ftps = FTP_TLS(self.hook.connection.host)
+        ftps.login(user=self.hook.connection.login, passwd=self.hook.connection.password)
+
+        yield ftps
+
+        ftps.quit()
+
+    def retrieve_file(self, ftps: FTP_TLS, file: FtpEntry) -> io.BufferedReader:
+        """
+        Opens the given file for download into a context as a python io (file like) object.
+        The underlying io object yields bytes objects.
+
+        Note:  Ideally, this method is used in combination with 'with' syntax in python, so that the underlying
+        data stream is closed after usage.  This is especially performant when code only needs to access a small
+        subset of the data.
+
+        eg:
+        ```
+        with box.open_file(file) as open_file:
+          for line in csv.reader(open_file):
+             break
+        ```
+
+        params:
+          ftps: an open, FTP_TLS connection
+          file: an FTP file listing
+        """
+        ftps.cwd(file.folder_path)
+
+        io_obj = io.BytesIO()
+        ftps.retrbinary(f"RETR {file.name}", io_obj.write)
```

Comment:
I, think this approach ends up buffering the entire download into memory, due to the push nature of the write.

There are a couple of solutions, and we should consider them, since file transfer of big files is not safe in buffered mode.

1.  Use `ntransfercmd` to retrieve a low level io and expected size value.  Using this combination, you can still construct a buffer reader, but one that reads in data in chunks as needed, rather than pushing the entire contents into memory immediately.  (https://pythontic.com/ftplib/ftp/ntransfercmd)
2.  Use a different library.  It looks like paramiko supports an open interface that supports streaming buffers https://docs.paramiko.org/en/stable/api/sftp.html#paramiko.sftp_client.SFTPClient.open     ie open("blah.txt", "r", 1024)

Reply:
Sorry, "buffered" readers are totes ok, what I meant by buffered here is "buffering the entire contents by pushing all of it immediately into a BytesIO", which is what retrbinary does unfortunately.

Buffered readers work by fillup to a certain size (a fixed buffer) on demand from reads, improving performance.